File: Top Interview Questions/215. Kth Largest Element in an Array.java
import java.util.Random;

class Solution {
    public int findKthLargest(int[] nums, int k) {
        return quickSelect(nums, 0, nums.length - 1, nums.length - k);
    }
    
    private void swap(int[] nums, int x, int y) {
        int t = nums[x];
        nums[x] = nums[y];
        nums[y] = t;
    }
    
    private int partition(int[] nums, int l, int r) {
        int p_index = (new Random().nextInt(r - l)) + l;
        
        swap(nums, p_index, r);
        int flag = l;
        int p = nums[r];
        for (int i = l; i < r; i++) {
            if (nums[i] < p) swap(nums, flag++, i);
        }
        
        swap(nums, flag, r);
        return flag;
    }
    
    private int quickSelect(int[] nums, int l, int r, int k) {
        if (l == r) return nums[l];
        
        int p = partition(nums, l, r);
        
        if (k == p) return nums[k];
        else if (k < p) 
            return quickSelect(nums, l, p - 1, k);
        
        return quickSelect(nums, p + 1, r, k);
    }
}

File: Top Interview Questions/Array/215. Kth Largest Element in an Array.py
class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        # qucik select
        def select(left, right, k_):
            if left == right:
                return nums[left]
            
            i = random.randint(left, right)
            i = partition(left, right, i)
            
            if k_ == i:
                return nums[i]
            elif k_ > i:
                # right part
                return select(i + 1, right, k_)
            else:
                # left part
                return select(left, i - 1, k_)
            
        def partition(left, right, i):
            nums[i], nums[right] = nums[right], nums[i]
            
            j = left
            for k in range(left, right):
                if nums[k] < nums[right]:
                    nums[j], nums[k] = nums[k], nums[j]
                    j += 1
                    
            nums[right], nums[j] = nums[j], nums[right]
            return j
        
        return select(0, len(nums) - 1, len(nums) - k)